python/ray/serve/batching.py

```diff
@@ -0,0 +1,275 @@
+        # in wait_for_batch.
+        if self.queue.qsize() == self.max_batch_size:
+            self.full_batch_event.set()
+
+    def qsize(self) -> int:
+        return self.queue.qsize()
+
+    async def wait_for_batch(self) -> List[Any]:
+        """Wait for batch respecting self.max_batch_size and self.timeout_s.
+
+        Returns a batch of up to self.max_batch_size items, waiting for up
+        to self.timeout_s for a full batch. After the timeout, returns as many
+        items as are ready.
+
+        Always returns a batch with at least one item - will block
+        indefinitely until an item comes in.
+        """
+        curr_timeout = self.timeout_s
+        batch = []
+        while len(batch) == 0:
+            loop_start = time.time()
+
+            # If the timeout is 0, wait for any item to be available on the
+            # queue.
+            if curr_timeout == 0:
+                batch.append(await self.queue.get())
+            # If the timeout is nonzero, wait for either the timeout to occur
+            # or the max batch size to be ready.
+            else:
+                try:
+                    await asyncio.wait_for(self.full_batch_event.wait(),
+                                           curr_timeout)
+                except asyncio.TimeoutError:
+                    pass
+
+            # Pull up to the max_batch_size requests off the queue.
+            while len(batch) < self.max_batch_size and not self.queue.empty():
+                batch.append(self.queue.get_nowait())
+
+            # Reset the event if there are fewer than max_batch_size requests
+            # in the queue.
+            if (self.queue.qsize() < self.max_batch_size
+                    and self.full_batch_event.is_set()):
+                self.full_batch_event.clear()
+
+            # Adjust the timeout based on the time spent in this iteration.
+            curr_timeout = max(0, curr_timeout - (time.time() - loop_start))
+
+        return batch
+
+    async def _handle_batches(self, func):
+        while True:
+            batch = await self.wait_for_batch()
+            assert len(batch) > 0
+            self_arg = batch[0][0]
+            args = [item[1] for item in batch]
+            futures = [item[2] for item in batch]
+
+            try:
+                # Method call.
+                if self_arg is not None:
+                    results = await func(self_arg, args)
+                # Normal function call.
+                else:
+                    results = await func(args)
+
+                if len(results) != len(batch):
+                    raise RayServeException(
+                        "Batched function doesn't preserve batch size. "
+                        f"The input list has length {len(batch)} but the "
+                        f"returned list has length {len(results)}.")
+            except Exception as e:
+                results = [_RaisedExceptionWrapper(e)] * len(batch)
```

Comment:
we actually don't need the wrapper here. you can just do `futures[i].set_exeception(e)`

Reply:
oh good call!